src/system_driver.cpp

```diff
@@ -553,6 +599,49 @@ void SystemDriver::ProjectH(ParGridFunction &h)
    return;
 }
 
+// This one requires that the deviatoric strain be converted from 5d rep to 6d
+// and have vol. contribution added.
+void SystemDriver::ProjectElasticStrains(ParGridFunction &estrain)
+{
+   if (mech_type == MechType::EXACMECH) {
+      std::string s_estrain = "elas_strain";
+      std::string s_rvol = "rel_vol";
+      auto qf_mapping = model->GetQFMapping();
+      auto espair = qf_mapping->find(s_estrain)->second;
+      auto rvpair = qf_mapping->find(s_rvol)->second;
+
+      const int e_offset = espair.first;
+      const int rv_offset = rvpair.first;
+
+      int _size = estrain.Size();
+      int nelems = _size / 6;
+
+      auto data_estrain = mfem::Reshape(estrain.HostReadWrite(), 6, nelems);
+      auto data_evec = mfem::Reshape(evec->HostReadWrite(), evec->GetVDim(), nelems);
+      // The below is outputting the full elastic strain in the crystal ref frame
+      // We'd only stored the 5d deviatoric elastic strain, so we need to convert
+      // it over to the 6d version and add in the volume elastic strain contribution.
+      for (int i = 0; i < nelems; i++) {
+         const double t1 = ecmech::sqr2i * data_evec(0 + e_offset, i);
+         const double t2 = ecmech::sqr6i * data_evec(1 + e_offset, i);
+         //
+         // Volume strain is ln(V^e_mean) term aka ln(relative volume)
+         // Our plastic deformation has a det(1) aka no change in volume change
+         const double elas_vol_strain = log(data_evec(rv_offset, i));
+         // In ECMech, they multiply the deviatoric portion by a 1 / det(V)^{1/3} term for Kirchoff stress calcs,
+         // so we should probably do that here as well.
+         const double a_inv = pow(data_evec(rv_offset, i), ecmech::onethird);
+         data_estrain(0, i) = a_inv * (t1 - t2) + elas_vol_strain; // 11'
+         data_estrain(1, i) = a_inv * (-t1 - t2) + elas_vol_strain ; // 22'
+         data_estrain(2, i) = a_inv * ecmech::sqr2b3 * data_evec(1 + e_offset, i) + elas_vol_strain; // 33'
+         data_estrain(3, i) = a_inv * ecmech::sqr2i * data_evec(4 + e_offset, i); // 23
+         data_estrain(4, i) = a_inv * ecmech::sqr2i * data_evec(3 + e_offset, i); // 31
+         data_estrain(5, i) = a_inv * ecmech::sqr2i * data_evec(2 + e_offset, i); // 12
```

Comment:
@nrbarton these are the relevant lines converting the deviatoric elastic strains and relative volume changes to the full elastic strain. I tried backing out all the relevant stuff based on what's being done in ECMech to calculate the Kirchoff stress and the code you mentioned in here: https://github.com/LLNL/ExaConstit/issues/40#issuecomment-980507035 . I believe I've got the right output now, but it's possible I'm still missing a scaling factor somewhere related to that `1/a` factor.

If you get the chance a second pair of eyes on this just to double check that everything is right would be nice.

Reply:
The details WITH the "a" factor are indeed confusing. I just spent a bunch of time digging through mspBEvpc_mspBEvpcEvptn_Impl.F90 and various other documentation. Given that a is close to 1 in most cases that we run, errors are not necessarily obvious in the results. I also looked at lines that read like `s_meas%Ee_vecds(1:TVEC) = a_V%ri * e_vecd` in elaw.F90 (and corresponding stuff in ECMech). As far as the formulation, some of the details are in the text near Equations 6 and 7 in [http://dx.doi.org/10.1063/1.1766762]. As far as I can remember and tell be looking at the code, the numbers stored in `data_evec` correspond to $\epsilon'$ (without the asterisk on it ...) so that you have to multiply by 1/a to get the actual strain measure and that is used in the evaluation of the Kirchhoff stress. But then in computing the stretch V there is another factor of a, so that a and 1/a cancel. See the comment on line 2234 of mspBEvpc_mspBEvpcEvptn_Impl.F90. In any case, I think your scaling by `a_inv` needs to go away -- but it is complicated enough that I could be wrong about that. 

I will also note that in mspBEvpc_mspBEvpcEvptn_Impl.F90 after rolling the deviator into the stretch the code rescales V so that the determinant comes out exactly to the value from the relative volume that is fed in; but that step is probably not necessary, and may not be the best approach in general. 

Also, the `a_inv` notation could be confusing, with people thinking the `_inv` means inverse so that it is 1/a rather than a that you are computing. 

And in your comments you should drop the prime from 11', 22', and 33', given that the thing you are computing here is not the deviator.

Reply:
@nrbarton thanks and yes getting this right is definitely a bit confusing at points... I think @darrencpagan is just using the elastic strains rather than the elastic left stretch tensor. However, I'll double check with him and see what's going on inside his script, so we make sure we're using the appropriate variables.

I'll fix the other portion of things as well.

Reply:
@nrbarton based on our other emails regarding this topic, you're right about the 1/a factor going away as the elastic strain stored within ECMech as a history variable is the one that corresponds to the I + \epsilon = V^e formulation rather than the a (I + \epsilon') \approx V^e formulation. So, we just need to add the volumetric contribution back, and then document that this is the I + \epsilon formulation of things.

Also, I'll try and finalize this over the next week.

Reply:
Updated this to correct things based on feedback we had earlier. I'm going to go ahead and merge this as the scaling factor is now corrected, and @darrencpagan has been able to successfully make use of the output for some light-up applications.